### PR TITLE
Ángel Vázquez - Image Comments Challenge

### PR DESCRIPTION
**Pull request by Angel5215**

### CODE REVIEW

File: EssentialApp/EssentialApp/ImageCommentCellPresentationAdapter.swift [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	var presenter: ImageCommentPresenter?
+	
+	init(comment: ImageComment) {
+		self.comment = comment
+	}
+	
+	func didRequestComment() {
+		presenter?.didLoadComment(comment)
+	}
+}
```
Comment (caiozullo) [inline EssentialApp/EssentialApp/ImageCommentCellPresentationAdapter.swift:23]: It seems like this adapter doesn't do anything - it just forwards the `model`. 

That's because there are two presenters handling the image comments logic. 

Instead, you could have a single presenter that maps the `ImageComment` model to `ImageCommentViewModel`.

Then, you could remove this adapter and pass the `ImageCommentViewModel` model directly to the cell controller via initializer injection.

File: EssentialApp/EssentialAppTests/FeedAcceptanceTests.swift [hunk 2]
```diff
@@ -7,7 +7,7 @@ import EssentialFeed
 import EssentialFeediOS
 @testable import EssentialApp
 
-class FeedAcceptanceTests: XCTestCase {
+class FeedAcceptanceTests: XCTestCase, AcceptanceTest {
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/FeedAcceptanceTests.swift:10]: Ideally, the new acceptance tests should be in the same file to make them easier to understand and maintain. The acceptance tests will check the overall flow in the app, so they should be together in this case as the feed and comments are part of the same flow. No need for abstraction.

File: EssentialApp/EssentialAppTests/Helpers/FeedImageCell+TestHelpers.swift [hunk 3]
```diff
@@ -10,6 +10,10 @@ extension FeedImageCell {
 		feedImageRetryButton.simulateTap()
 	}
 	
+	func simulateTapAction() {
+		feedImageButton.simulateTap()
+	}
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/Helpers/FeedImageCell+TestHelpers.swift:15]: Avoid adding buttons for cell selection. Instead, use the table view delegate to handle cell selection - it's easier and provides better UX.

File: EssentialApp/EssentialAppTests/FeedUIIntegrationTests.swift [hunk 4]
```diff
@@ -210,6 +210,62 @@ final class FeedUIIntegrationTests: XCTestCase {
 		XCTAssertEqual(view1?.isShowingRetryAction, false, "Expected no retry action for second view on retry")
 	}
 	
+	func test_feedImageViewImageButton_isNotVisibleUntilImageLoadsSuccessfully() {
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/FeedUIIntegrationTests.swift:213]: As described in my previous comment, there shouldn't be complex logic about cell selection. It should always be selectable through the table view delegate method.

File: EssentialApp/EssentialAppTests/Helpers/ImageCommentsIntegrationTests+Assertions.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		guard let cell = view as? ImageCommentCell else {
+			return XCTFail("Expected \(ImageCommentCell.self) instance, got \(String(describing: self)) instead", file: file, line: line)
+		}
+		
+		XCTAssertEqual(cell.authorText, pair.comment.author, "Expected author to be \(pair.comment.author) for comment view at index \(index)", file: file, line: line)
+		
+		XCTAssertEqual(cell.messageText, pair.comment.message, "Expected message to be \(pair.comment.message) for comment view at index \(index)", file: file, line: line)
+		
+		XCTAssertEqual(cell.creationDateText, pair.relativeDate, "Expected relative date to be \(pair.relativeDate) for comment view at index \(index)", file: file, line: line)
+	}
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/Helpers/ImageCommentsIntegrationTests+Assertions.swift:36]: `XCTAssertEqual` already logs the expected value, so the assertion messages could be shorter. 

For example, just referencing the property (e.g., authorText) and the cell index:

```
XCTAssertEqual(cell.authorText, pair.comment.author, "authorText at index (\(index))", file: file, line: line)
```

This would log:

`XCTAssertEqual failed: ("bob") is not equal to ("alice") - authorText at index 0`

File: EssentialApp/EssentialAppTests/ImageCommentsIntegrationTests.swift [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	
+	func test_loadCommentsCompletion_rendersSuccessfullyLoadedComments() {
+		let staticDate = makeDateFromTimestamp(1_605_868_247, description: "2020-11-20 10:30:47 +0000")
+		
+		let pair0 = makeCommentDatePair(
+			message: "a message",
+			creationDate: makeDateFromTimestamp(1_605_860_313, description: "2020-11-20 08:18:33 +0000"),
+			author: "an author",
+			expectedRelativeDate: "2 hours ago"
+		)
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/ImageCommentsIntegrationTests.swift:63]: Formatting logic should be tested in the presenter tests. Otherwise, this test would fail if the presentation logic changes - you'd have to update two files: the presenter tests and the integrations tests.

 The integration tests shouldn't fail if the presentation logic changes.

Remove every logic about formatting from the integration tests - you can use the presenter method directly to create the values here.

File: EssentialApp/EssentialAppTests/ImageCommentsIntegrationTests.swift [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+		let loader = LoaderSpy()
+		let sut = ImageCommentsUIComposer.imageCommentsComposedWith(url: url, currentDate: currentDate, loader: loader)
+		trackForMemoryLeaks(loader, file: file, line: line)
+		trackForMemoryLeaks(sut, file: file, line: line)
+		return (sut, loader)
+	}
+	
+	private func makeDateFromTimestamp(_ timestamp: TimeInterval, description: String, file: StaticString = #file, line: UInt = #line) -> Date {
+		let date = Date(timeIntervalSince1970: timestamp)
+		XCTAssertEqual(date.description, description, file: file, line: line)
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/ImageCommentsIntegrationTests.swift:223]: `date.description` is not a precise evaluation of a `date` instance.

You shouldn't need this assertion - remove it.

File: EssentialApp/EssentialAppTests/SceneDelegateTests.swift [hunk 8]
```diff
@@ -32,5 +32,24 @@ class SceneDelegateTests: XCTestCase {
 		XCTAssertNotNil(rootNavigation, "Expected a navigation controller as root, got \(String(describing: root)) instead")
 		XCTAssertTrue(topController is FeedViewController, "Expected a feed controller as top view controller, got \(String(describing: topController)) instead")
 	}
+
+	func test_handleImageID_pushesImageCommentsViewController() {
+		let sut = SceneDelegate()
+		sut.window = UIWindow()
+		sut.configureWindow()
+		
+		sut.handleImageID(UUID().uuidString)
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/SceneDelegateTests.swift:41]: `handleImageID` should be private - this navigation logic should be tested through the acceptance tests.

File: EssentialFeed/EssentialFeed/Feed Presentation/FeedImageViewModel.swift [hunk 9]
```diff
@@ -12,4 +12,8 @@ public struct FeedImageViewModel<Image> {
 	public var hasLocation: Bool {
 		return location != nil
 	}
+	
+	public var shouldAllowImageTap: Bool {
+		return !isLoading && !shouldRetry
+	}
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeed/Feed Presentation/FeedImageViewModel.swift:18]: This property is not needed. The cell should always be tappable.

File: EssentialFeed/EssentialFeed/Image Comment Feature/ImageCommentLoader.swift [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import Foundation
+
+public protocol ImageCommentLoaderTask {
+	func cancel()
+}
+
+public protocol ImageCommentLoader {
+	typealias Result = Swift.Result<[ImageComment], Error>
+	
+	func load(from url: URL, completion: @escaping (Result) -> Void) -> ImageCommentLoaderTask
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeed/Image Comment Feature/ImageCommentLoader.swift:18]: `URL` is an implementation detail of the `Remote` loader. Not *every* implementation should need a URL.

The protocol should not expose implementation details. Otherwise, it's not a good abstraction.

The URL should be injected into the concrete Remote loader implementation via initializer injection instead.

```suggestion
	func load(completion: @escaping (Result) -> Void) -> ImageCommentLoaderTask
```

File: EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentPresenter.swift [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	}
+	
+	public func didLoadComment(_ comment: ImageComment) {
+		commentView.display(
+			ImageCommentViewModel(
+				author: comment.author,
+				message: comment.message,
+				creationDate: formatRelativeDate(for: comment.creationDate)
+			)
+		)
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentPresenter.swift:43]: This logic should be in the `ImageCommentsListPresenter` - so you can remove this presenter class.

The code will be much simpler to develop, test, and compose with a single presenter.

File: EssentialFeed/EssentialFeedEndToEndTests/EssentialFeedEndToEndTests.swift [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+//  EssentialFeedAPIEndToEndTests
+//
+//  Created by Ángel Vázquez on 19/03/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import EssentialFeed
+import XCTest
+
+class EssentialFeedAPIEndToEndTests: XCTestCase {
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedEndToEndTests/EssentialFeedEndToEndTests.swift:12]: End-to-end tests are not part of this challenge. Please, delete the tests and the target.

File: EssentialFeed/EssentialFeedTests/Image Comments API /LoadImageCommentsFromRemoteUseCaseTests.swift [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+		})
+	}
+	
+	func test_load_deliversInvalidDataErrorOnNon2xxHTTPResponse() {
+		let (sut, client) = makeSUT()
+		let samples = [199, 300, 400, 500]
+		
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: failure(.invalidData), when: {
+				client.complete(withStatusCode: code, data: anyData(), at: index)
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedTests/Image Comments API /LoadImageCommentsFromRemoteUseCaseTests.swift:53]: `anyData` is not a valid response for this endpoint based on API the contract. It should use a valid response to be able to test the `status code` logic.

It could be an empty JSON list: 

```suggestion
				client.complete(withStatusCode: code, data: makeCommentsJSON([]), at: index)
```

File: EssentialFeed/EssentialFeedTests/Image Comments API /LoadImageCommentsFromRemoteUseCaseTests.swift [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	
+	func test_load_deliversErrorOn200HTTPResponseWithInvalidJSON() {
+		let (sut, client) = makeSUT()
+		
+		expect(sut, toCompleteWith: failure(.invalidData), when: {
+			let invalidJSON = Data("invalid json".utf8)
+			client.complete(withStatusCode: 200, data: invalidJSON)
+		})
+	}
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedTests/Image Comments API /LoadImageCommentsFromRemoteUseCaseTests.swift:65]: All the test methods in this class should check for 2xx status code with 2xx samples (not just 200).

Make sure to update all tests, including their names.

File: EssentialFeed/EssentialFeedTests/Image Comments Presentation/ImageCommentPresenterTests.swift [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+			(comment(date: dateFromTimestamp(1_605_860_313, description: "2020-11-20 08:18:33 +0000")), "2 hours ago"),
+			(comment(date: dateFromTimestamp(1_605_713_544, description: "2020-11-18 15:32:24 +0000")), "1 day ago"),
+			(comment(date: dateFromTimestamp(1_604_571_429, description: "2020-11-05 10:17:09 +0000")), "2 weeks ago"),
+			(comment(date: dateFromTimestamp(1_602_510_149, description: "2020-10-12 13:42:29 +0000")), "1 month ago"),
+			(comment(date: dateFromTimestamp(1_488_240_000, description: "2017-02-28 00:00:00 +0000")), "3 years ago")
+		]
+		
+		samples.enumerated().forEach { index, pair in
+			let (comment, relativeDate) = pair
+			let (sut, view) = makeSUT(currentDate: { staticDate })
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedTests/Image Comments Presentation/ImageCommentPresenterTests.swift:31]: Make sure the test is also environment-independent. You could inject an English locale (e.g., `en_US_POSIX`), so the test will not rely on global environment details that might change. Otherwise, the test may fail if running in a different locale because it's expecting the English relative date strings, e.g., `1 minute ago`.

File: EssentialFeed/EssentialFeedTests/Image Comments Presentation/ImageCommentPresenterTests.swift [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+			
+			sut.didLoadComment(comment)
+			
+			XCTAssertEqual(
+				view.messages,
+				[.display(author: comment.author, message: comment.message, relativeDate: relativeDate)],
+				"Expected comment at index \(index) to have author '\(comment.author)', message '\(comment.message)' and relative date '\(relativeDate)'"
+			)
+		}
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedTests/Image Comments Presentation/ImageCommentPresenterTests.swift:40]: The message is not needed since `XCTAssertEqual` will log the values.

 ```suggestion
			XCTAssertEqual(
				view.messages,
				[.display(author: comment.author, message: comment.message, relativeDate: relativeDate)]
			)
		}
```

File: EssentialFeed/EssentialFeedTests/Image Comments Presentation/ImageCommentPresenterTests.swift [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+		let view = ViewSpy()
+		let sut = ImageCommentPresenter(currentDate: currentDate, commentView: view)
+		trackForMemoryLeaks(view, file: file, line: line)
+		trackForMemoryLeaks(sut, file: file, line: line)
+		return (sut, view)
+	}
+	
+	private func dateFromTimestamp(_ timestamp: TimeInterval, description: String, file: StaticString = #file, line: UInt = #line) -> Date {
+		let date = Date(timeIntervalSince1970: timestamp)
+		XCTAssertEqual(date.description, description, file: file, line: line)
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedTests/Image Comments Presentation/ImageCommentPresenterTests.swift:55]: The assertion is not needed (see my other comment). Remove it.

File: EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsCellController.swift [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+public protocol ImageCommentCellControllerDelegate {
+	func didRequestComment()
+}
+
+public final class ImageCommentsCellController: ImageCommentView {
+	private let delegate: ImageCommentCellControllerDelegate
+	private var cell: ImageCommentCell?
+	
+	public init(delegate: ImageCommentCellControllerDelegate) {
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsCellController.swift:20]: The model data is available at creation time, so it shouldn't receive a delegate. It should receive the view model instead. This way, you don't need to capture a reference to the cell in a property.

File: EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+		sut.didStartLoadingComments()
+		
+		XCTAssertEqual(view.messages, [.display(errorMessage: .none), .display(isLoading: true)])
+	}
+	
+	func test_didFinishLoadingCommentsWithComments_stopsLoadingAndDisplaysComments() {
+		let (sut, view) = makeSUT()
+		let comments = makeComments()
+		
+		sut.didFinishLoadingComments(with: comments)
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift:31]: This presenter should do all the mapping logic, including date formatting. 

There should be only one image comments presenter. 

Move the logic to a single presenter and move the formatting tests here.

File: EssentialFeed/EssentialFeediOSTests/ImageCommentsSnapshotTests.swift [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+		assert(snapshot: sut.snapshot(for: .iPhone8(style: .dark)), named: "EMPTY_IMAGE_COMMENTS_dark")
+	}
+	
+	func test_commentsListWithContent() {
+		let sut = makeSUT()
+		
+		sut.display(commentsListWithContent())
+		
+		assert(snapshot: sut.snapshot(for: .iPhone8(style: .light)), named: "IMAGE_COMMENTS_WITH_CONTENT_light")
+		assert(snapshot: sut.snapshot(for: .iPhone8(style: .dark)), named: "IMAGE_COMMENTS_WITH_CONTENT_dark")
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOSTests/ImageCommentsSnapshotTests.swift:29]: Missing a snapshot to check Dynamic Type works (e.g., larger fonts with extraExtraExtraLarge content size)

File: EssentialFeed/EssentialFeediOSTests/ImageCommentsSnapshotTests.swift [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	
+	func test_commentsListWithErrorMessage() {
+		let sut = makeSUT()
+		
+		sut.refreshController?.display(ImageCommentsErrorViewModel(message: "This is a\nmulti-line\nerror message"))
+		
+		assert(snapshot: sut.snapshot(for: .iPhone8(style: .light)), named: "IMAGE_COMMENTS_WITH_ERROR_MESSAGE_light")
+		assert(snapshot: sut.snapshot(for: .iPhone8(style: .dark)), named: "IMAGE_COMMENTS_WITH_ERROR_MESSAGE_dark")
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOSTests/ImageCommentsSnapshotTests.swift:38]: Missing a snapshot to check Dynamic Type works (e.g., larger fonts with extraExtraExtraLarge content size)

**Review by caiozullo — COMMENTED** [review]
Hi @Angel5215, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. Thank you 🙂

**Comment by Angel5215** [comment]
Hi @caiozullo, thanks for your feedback! 

I've addressed your comments, could you review again? Thanks!

File: EssentialApp/EssentialAppTests/Helpers/FeedViewController+TestHelpers.swift [hunk 22]
```diff
@@ -26,6 +26,13 @@ extension FeedViewController {
 		return view
 	}
 	
+	func simulateTapOnFeedImageView(at row: Int) {
+		simulateFeedImageViewVisible(at: row)
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/Helpers/FeedViewController+TestHelpers.swift:30]: Calling `simulateFeedImageViewVisible` is not needed in this method - you can remove it.

```suggestion
```

File: EssentialApp/EssentialAppTests/FeedUIIntegrationTests.swift [hunk 23]
```diff
@@ -210,6 +210,17 @@ final class FeedUIIntegrationTests: XCTestCase {
...
 	
+	func test_feedImageView_forwardsFeedImageIDOnTapWhenImageHasLoadSuccessfully() {
+		let image = makeImage()
+		var capturedImages = [FeedImage]()
+		let (sut, loader) = makeSUT(imageHandler: { capturedImages.append($0) })
+		
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoading(with: [image])
+		sut.simulateTapOnFeedImageView(at: 0)
+		XCTAssertEqual(capturedImages, [image], "Expected tapped image to be forwarded without changes")
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/FeedUIIntegrationTests.swift:221]: Missing pre-condition assertion (no image should have been selected). 

```suggestion
		sut.loadViewIfNeeded()
		loader.completeFeedLoading(with: [image])
		XCTAssertEqual(capturedImages, [])
		
		sut.simulateTapOnFeedImageView(at: 0)
		XCTAssertEqual(capturedImages, [image], "Expected tapped image to be forwarded without changes")
```

Also, you should add at least 2 images and select both at different indexes to increase coverage. One is not enough.

File: EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentsListPresenter.swift [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+import Foundation
+
+// MARK: - ViewModels
+
+public struct ImageCommentsLoadingViewModel {
+	public let isLoading: Bool
+}
+
+public struct ImageCommentsListViewModel {
+	public let comments: [ImageComment]
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentsListPresenter.swift:18]: The type in the comments array should be `ImageCommentViewModel` - the presenter should map `ImageComment` to `ImageCommentViewModel`.

This way, the code is much simpler because clients don't need to map the `ImageComment`s later.

File: EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+		trackForMemoryLeaks(view, file: file, line: line)
+		trackForMemoryLeaks(sut, file: file, line: line)
+		return (sut, view)
+	}
+	
+	private func dateFromTimestamp(_ timestamp: TimeInterval, description: String, file: StaticString = #file, line: UInt = #line) -> Date {
+		Date(timeIntervalSince1970: timestamp)
+	}
+	
+	func makeComments() -> [ImageComment] {
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift:79]: This method should be private.

File: EssentialApp/EssentialAppTests/ImageCommentsIntegrationTests.swift [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+		let loader = LoaderSpy()
+		let sut = ImageCommentsUIComposer.imageCommentsComposedWith(url: url, currentDate: currentDate, loader: loader)
+		trackForMemoryLeaks(loader, file: file, line: line)
+		trackForMemoryLeaks(sut, file: file, line: line)
+		return (sut, loader)
+	}
+	
+	private func makeDateFromTimestamp(_ timestamp: TimeInterval, description: String, file: StaticString = #file, line: UInt = #line) -> Date {
+		Date(timeIntervalSince1970: timestamp)
+	}
```
Comment (caiozullo) [inline EssentialApp/EssentialAppTests/ImageCommentsIntegrationTests.swift:218]: The parameters passed to this function are never used. You should remove this function and just create the date directly where needed.

File: EssentialFeed/EssentialFeedTests/Image Comments API /LoadImageCommentsFromRemoteUseCaseTests.swift [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: failure(.invalidData), when: {
+				client.complete(withStatusCode: code, data: makeCommentsJSON([]), at: index)
+			})
+		}
+	}
+	
+	func test_load_deliversErrorOn2xxHTTPResponseWithInvalidJSON() {
+		let (sut, client) = makeSUT(url: anyURL())
+		let samples = [200, 201, 202, 206, 207, 226]
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeedTests/Image Comments API /LoadImageCommentsFromRemoteUseCaseTests.swift:60]: It's important to also add at least `200, 201, and 299` to the samples to test the edges and avoid off-by-one errors.

Please, updated all appropriate tests to include 299.

File: EssentialFeed/EssentialFeediOS/Feed UI/Views/FeedImageCell.swift [hunk 28]
```diff
@@ -13,6 +13,7 @@ public final class FeedImageCell: UITableViewCell {
 	@IBOutlet private(set) public var descriptionLabel: UILabel!
 	
 	var onRetry: (() -> Void)?
+	var onTap: (() -> Void)?
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOS/Feed UI/Views/FeedImageCell.swift:16]: This property is not used. Remove it.

File: EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+		let view = ViewSpy()
+		let sut = ImageCommentsListPresenter(loadingView: view, commentsView: view, errorView: view)
+		trackForMemoryLeaks(view, file: file, line: line)
+		trackForMemoryLeaks(sut, file: file, line: line)
+		return (sut, view)
+	}
+	
+	private func dateFromTimestamp(_ timestamp: TimeInterval, description: String, file: StaticString = #file, line: UInt = #line) -> Date {
+		Date(timeIntervalSince1970: timestamp)
+	}
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift:77]: The parameters passed to this function are never used. You should remove this function and just create the date directly where needed.

File: EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+			(comment(date: dateFromTimestamp(1_605_713_544, description: "2020-11-18 15:32:24 +0000")), "1 day ago"),
+			(comment(date: dateFromTimestamp(1_604_571_429, description: "2020-11-05 10:17:09 +0000")), "2 weeks ago"),
+			(comment(date: dateFromTimestamp(1_602_510_149, description: "2020-10-12 13:42:29 +0000")), "1 month ago"),
+			(comment(date: dateFromTimestamp(1_488_240_000, description: "2017-02-28 00:00:00 +0000")), "3 years ago")
+		]
+		
+		samples.enumerated().forEach { index, pair in
+			let (comment, relativeDate) = pair
+			
+			let viewModel = ImageCommentsListPresenter.viewModel(for: comment, currentDate: { staticDate })
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeediOS/Image Comments UI/ImageCommentsListPresenterTests.swift:57]: Make sure the test is environment-independent. You could inject an English locale (e.g., `en_US_POSIX`), so the test will not rely on global environment details that might change. Otherwise, the test may fail if running in a different locale because it's expecting the English relative date strings, e.g., `1 minute ago`.

**Review by caiozullo — COMMENTED** [review]
Hi @Angel5215, 

Congratulations, you've passed the challenge 🥇

✅ Display a list of comments when the user taps on an image in the feed. 
	
	✅ Show a loading spinner while loading the comments.
		
	✅ If it fails to load: Show an error message.
		
	✅ If it loads successfully: Show all loaded comments in the order they were returned by the remote API.

	✅ Start loading automatically when the user navigates to the screen.

	✅ The user should also be able to reload the comments manually (Pull-to-refresh).

	✅ The user should have a back button to return to the feed screen.

	✅ Cancel any running comments API requests when the user navigates back.

✅ The comments screen layout should match the UI specs.
	
	✅ Present the comment date using relative date formatting, e.g., "1 day ago."

	✅ The title should be localized in all languages supported in the project.

	✅ Light and Dark Mode
	
	✅ Dynamic Fonts

✅ Write tests to validate your implementation:
	
	✅ Unit tests

	✅ Snapshot tests

	✅ Integration tests

	✅ Acceptance tests

I left some final review comments. Feel free to push changes.

Great job!

---

I highly recommend you watch the live sessions now to learn how to eliminate duplication in the code without breaking modularity.

Keep up the great work 🚀

File: EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentsListPresenter.swift [hunk 31]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	
+	public func didFinishLoadingComments(with error: Error) {
+		loadingView.display(ImageCommentsLoadingViewModel(isLoading: false))
+		errorView.display(ImageCommentsErrorViewModel(message: Localized.ImageComments.errorMessage))
+	}
+}
+
+public final class ImageCommentViewModelMapper {
+	public static func map(_ comments: [ImageComment], currentDate: () -> Date, locale: Locale) -> [ImageCommentViewModel] {
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentsListPresenter.swift:90]: Since this static `map` method is public, you should also write tests for it because clients might use it directly.

File: EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentsListPresenter.swift [hunk 32]
```diff
@@ -0,0 +1,107 @@
...
+		}
+	}
+}
+
+private extension Date {
+	func dateString(relativeTo date: Date, locale: Locale) -> String {
+		let formatter = RelativeDateTimeFormatter()
+		formatter.locale = locale
+		return formatter.localizedString(for: self, relativeTo: date)
+	}
```
Comment (caiozullo) [inline EssentialFeed/EssentialFeed/Image Comments Presentation/ImageCommentsListPresenter.swift:106]: This method is called in a loop (map). It'll be called for each element in the array and create a new `RelativeDateTimeFormatter` instance for each element.

The problem is: RelativeDateTimeFormatter is an expensive object to create. Ideally, it shouldn't be created in a loop for each element. It should be created once and reuse in the loop.
